FlowCrypt/Controllers/SearchViewController.swift

```diff
@@ -0,0 +1,321 @@
+//  Copyright © 2019 FlowCrypt Limited. All rights reserved.
+//
+
+import AsyncDisplayKit
+
+final class SearchViewController: ASViewController<TableNode> {
+    enum State {
+        enum FetchedUpdates {
+            case added(Int), removed(Int)
+        }
+        case idle, startFetching, empty, fetched([MCOIMAPMessage], FetchedUpdates?), error(String)
+        
+        var messages: [MCOIMAPMessage] {
+            guard case let .fetched(messages, _) = self else { return [] }
+            return messages
+        }
+    }
+    
+    private var state: State = .idle {
+        didSet { updateState() }
+    }
+    
+    private let messageProvider: SearchResultsProvider
+    private var searchTask: DispatchWorkItem?
+    
+    private let searchController = UISearchController(searchResultsController: nil)
+    private let folderPath: String
+    private var searchedExpression: String = ""
+    
+    init(
+        messageProvider: SearchResultsProvider = Imap.shared,
+        folderPath: String
+    ) {
+        self.messageProvider = messageProvider
+        self.folderPath = folderPath
+        super.init(node: TableNode())
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupUI()
+        setupSearch()
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        if case .idle = state {
+            searchController.isActive = true
+        }
+    }
+}
+
+// MARK: - Setup
+
+extension SearchViewController {
+    private func setupUI() {
+        view.backgroundColor = .white
+        title = "search_title".localized
+        node.delegate = self
+        node.dataSource = self
+    }
+
+    private func setupSearch() {
+        searchController.do {
+            $0.delegate = self
+            $0.searchResultsUpdater = self
+            $0.hidesNavigationBarDuringPresentation = false
+            $0.searchBar.tintColor = .white
+            $0.searchBar.setImage( #imageLiteral(resourceName: "search_icn").tinted(.white), for: .search, state: .normal)
+            $0.searchBar.setImage( #imageLiteral(resourceName: "cancel.png").tinted(.white), for: .clear, state: .normal)
+            $0.searchBar.delegate = self
+            $0.searchBar.textField?.textColor = .white
+            if #available(iOS 12, *) {
+                $0.searchBar.textField?.backgroundColor = .main
+            }
+        }
+        update(searchController: searchController)
+        definesPresentationContext = true
+        navigationItem.titleView = searchController.searchBar
+    }
+ 
+    @objc private func handleInfoTap() {
+        #warning("ToDo")
+        showToast("Email us at human@flowcrypt.com")
+    }
+}
+
+// MARK: - MessageHandlerViewConroller
+
+extension SearchViewController: MessageHandlerViewConroller {
```

Comment:
How is this used? The search screen does not have any delete or markAsRead buttons. Isn't this duplicating code from MsgView?

Reply:
I will fix this bug in the video. 
No, it isn't code duplicate. MessageHandlerViewConroller handles read and remove actions after you open the message. 
For example you have searched for google -> received 10 results -> open first message -> delete it. In this case, it will remove this message from the list. Same with reading.

Reply:
`extension InboxViewController: MessageHandlerViewConroller {` contains nearly identical code. Why extend the same functionality two times? Why not just inherit it? I do see this as code duplication.

Reply:
Why it's extneded twice? There is a protocol MessageHandlerViewConroller which have 2 functions 
```
func openMessageIfPossible(with message: MCOIMAPMessage, path: String)
func handleMessage(operation: MsgViewController.MessageAction, message: MCOIMAPMessage)
```

And I made a protocol with a default implementation for `openMessageIfPossible` function
```
extension MessageHandlerViewConroller where Self: UIViewController {
    func openMessageIfPossible(with message: MCOIMAPMessage, path: String) {
  
    }
}
```
but because datasource is different in every viewcontroller - there is no way to make default implementation for

`func handleMessage(operation: MsgViewController.MessageAction, message: MCOIMAPMessage)`

Reply:
I see it a little better now. But `handleMessage` seems to be the same for both implementations. Let me take a look.